cloud-controller-manager/do/kubernetes_lb_validator_test.go

```diff
@@ -0,0 +1,401 @@
+	panic("not supported")
+}
+
+func (f *fakeKubernetesService) CreateNodePool(ctx context.Context, clusterID string, req *godo.KubernetesNodePoolCreateRequest) (*godo.KubernetesNodePool, *godo.Response, error) {
+	panic("not supported")
+}
+
+func (f *fakeKubernetesService) GetNodePool(ctx context.Context, clusterID, poolID string) (*godo.KubernetesNodePool, *godo.Response, error) {
+	panic("not supported")
+}
+
+func (f *fakeKubernetesService) UpdateNodePool(ctx context.Context, clusterID, poolID string, req *godo.KubernetesNodePoolUpdateRequest) (*godo.KubernetesNodePool, *godo.Response, error) {
+	panic("not supported")
+}
+
+func (f *fakeKubernetesService) RecycleNodePoolNodes(ctx context.Context, clusterID, poolID string, req *godo.KubernetesNodePoolRecycleNodesRequest) (*godo.Response, error) {
+	panic("not supported")
+}
+
+func (f *fakeKubernetesService) DeleteNodePool(ctx context.Context, clusterID, poolID string) (*godo.Response, error) {
+	panic("not supported")
+}
+
+func (f *fakeKubernetesService) DeleteNode(ctx context.Context, clusterID, poolID, nodeID string, req *godo.KubernetesNodeDeleteRequest) (*godo.Response, error) {
+	panic("not supported")
+}
+
+func (f *fakeKubernetesService) GetOptions(ctx2 context.Context) (*godo.KubernetesOptions, *godo.Response, error) {
+	panic("not supported")
+}
+
+func (f *fakeKubernetesService) AddRegistry(ctx context.Context, req *godo.KubernetesClusterRegistryRequest) (*godo.Response, error) {
+	panic("not supported")
+}
+
+func (f *fakeKubernetesService) RemoveRegistry(ctx context.Context, req *godo.KubernetesClusterRegistryRequest) (*godo.Response, error) {
+	panic("not supported")
+}
+
+func (f *fakeKubernetesService) RunClusterlint(ctx context.Context, clusterID string, req *godo.KubernetesRunClusterlintRequest) (string, *godo.Response, error) {
+	panic("not supported")
+}
+
+func (f *fakeKubernetesService) GetClusterlintResults(ctx context.Context, clusterID string, req *godo.KubernetesGetClusterlintRequest) ([]*godo.ClusterlintDiagnostic, *godo.Response, error) {
+	panic("not supported")
+}
+
+func (f *fakeKubernetesService) ListNodePools(ctx context.Context, clusterID string, opts *godo.ListOptions) ([]*godo.KubernetesNodePool, *godo.Response, error) {
+	cluster := fakeCluster()
+	return cluster.NodePools, newFakeOKResponse(), nil
+}
+
+type fakeRegionsService struct {
+	listFn func(context.Context, *godo.ListOptions) ([]godo.Region, *godo.Response, error)
+}
+
+func (f *fakeRegionsService) List(ctx context.Context, listOpts *godo.ListOptions) ([]godo.Region, *godo.Response, error) {
+	return f.listFn(ctx, listOpts)
+}
+
+func Test_Handle(t *testing.T) {
+	//cluster := fakeCluster()
+	testcases := []struct {
+		name            string
+		req             admission.Request
+		oldObject       runtime.RawExtension
+		gCLient         *godo.Client
+		lbRequest       *godo.LoadBalancerRequest
+		expectedAllowed bool
+		expectedReason  string
```

Comment:
If the reason is part of every denied and errored validation response, then please specify one everywhere where `expectedAllowed` is false: this ensures we are testing the right code path and not fall prey to a false positive.

It's fine to use an identifying reason substring (as opposed to exact string) for that to ease matching.

Reply:
Dropping this because it's handled by checking for error types as well as additional coverage will be added when individual cases for different annotations is added

Reply:
Do we check the error types anywhere? I don't see it right now.

Also, even if we did I think we still want to test for exact or substring reason matching: otherwise, we cannot distinguish between different errors from the same type (e.g., create denied vs update denied) which is a common source for false positives in unit tests.

If we do this, then we can also drop `expectedAllow` in favor of `expectedError` (or `expectedErrorMessage`) where a nil/empty value implies that allowance should have been granted.